src/osd/PrimaryLogPG.cc

```diff
@@ -5518,7 +5518,12 @@ int PrimaryLogPG::do_read(OpContext *ctx, OSDOp& osd_op) {
     }
     if (r >= 0)
       op.extent.length = r;
-    else {
+    else if (r == -EAGAIN) {
+      // EAGAIN should not change the length of extent or count the read op.
+      dout(10) << " read got " << r << " / " << op.extent.length
+              << " bytes from obj " << soid << ". try again." << dendl;
+      return -EAGAIN;
```

Comment:
> can we just assign `r` to `result` here, and fix the `// XXX` two lines below in another commit, which updates `delta_stats` only if `result >= 0`.

Yes, we can just assign `r` to `result` here. But the one who designed this may want to record the read with `num_rd` whether the read failed or not, when the read op is completed. If we deal with this, the modification will be much more complicated. 
Just let it is may be better. Thanks for your review.

Reply:
> the one who designed this may want to record the read with num_rd whether the read failed or not, when the read op is completed. 

i am quoting the comment here

>   // XXX the op.extent.length is the requested length for async read
>  // On error this length is changed to 0 after the error comes back.

@dzafman do you think we should take failed read into account?

>  the modification will be much more complicated.

i doubt that. it'd add one more `if` clause which is far from "complicated".

Reply:
#25687